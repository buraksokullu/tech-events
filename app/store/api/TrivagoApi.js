import TrivagoLocalService from 'Services/index';
import {
  getEventsLoading,
  getEventsSuccess,
  getEventsFailure,
  getCitiesLoading,
  getCitiesSuccess,
  getCitiesFailure
} from '../actions/index';

export const getEvents = () => async dispatch => {
  await new Promise((resolve, reject) => {
    dispatch(getEventsLoading(true));
    TrivagoLocalService.getEvents()
      .then(response => {
        dispatch(getEventsSuccess(response.data));
        resolve();
      })
      .catch(error => {
        dispatch(getEventsFailure(error));
        reject();
      });
  });
};

export const getCities = () => async dispatch => {
  await new Promise((resolve, reject) => {
    dispatch(getCitiesLoading(true));
    TrivagoLocalService.getCities()
      .then(response => {
        dispatch(getCitiesSuccess(response.data));
        resolve(response.data);
      })
      .catch(error => {
        dispatch(getCitiesFailure(error));
        reject();
      });
  });
};
