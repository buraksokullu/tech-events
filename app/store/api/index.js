import { getCities, getEvents } from './trivagoApi';

export { getCities, getEvents };
