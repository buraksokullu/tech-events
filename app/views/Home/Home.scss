.bodyHolder {
  display: flex;
  flex-direction: row;
  margin-top: 20px;
  max-width: 1200px;
  width: 100%;
  margin-left: auto;
  margin-right: auto;
  box-sizing: border-box;

  @media only screen and (max-width: 1024px) {
    flex-direction: column;
    margin-top: 16px;
    padding: 0 12px;
  }
}
.childrenHolder {
  flex: 1;
}
