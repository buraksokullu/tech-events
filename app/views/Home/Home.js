import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import { getEvents, getCities } from 'Store/api/index';
import EventTypeEnums from 'Model/index';
import { eventsSelector, citiesSelector } from 'Store/selectors/index';
import { toastr } from 'react-redux-toastr';

import Button from 'Components/FormFields/Button/Button';
import Modal from 'Components/Modal/Modal';
import Header from 'Components/Layout/Header/Header';
import Aside from 'Components/Layout/Aside/Aside';
import Events from '../Events/Events';

import s from './Home.scss';

export class Home extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showModal: false,
      eventName: '',
      registeredEvents: [],
      citiesState: [],
      selectedEvents: [],
      allEvents: []
    };
  }

  componentDidMount() {
    const { getCities } = this.props;
    getCities();
    this.getEvents();
  }

  getEvents = async () => {
    return new Promise(resolve => {
      const { getEvents } = this.props;

      getEvents().then(() => {
        this.setData();
        resolve();
      });
    });
  };

  setData = () => {
    const { cities, events } = this.props;
    const selectedEventIdsStore = JSON.parse(localStorage.getItem('selectedEvents'));

    const newArr = events.map(v => ({
      ...v,
      isSelected: selectedEventIdsStore ? !!selectedEventIdsStore.includes(v.id) : false
    }));

    this.setState({
      allEvents: newArr,
      selectedEvents: newArr,
      citiesState: cities,
      registeredEvents: selectedEventIdsStore || []
    });
  };

  filterByName = data => {
    const { selectedEvents, allEvents } = this.state;
    const filteredEvents = data
      ? selectedEvents.filter(ev => ev.name.toLowerCase().includes(data.toLowerCase()))
      : allEvents;

    this.setState({
      selectedEvents: filteredEvents
    });
  };

  filterFreeEvents = data => {
    const { selectedEvents, allEvents } = this.state;

    const filteredEvents = data ? selectedEvents.filter(ev => ev.isFree === true) : allEvents;

    this.setState({
      selectedEvents: filteredEvents
    });
  };

  toggleModal = (id, eventName) => {
    this.setState(prevState => ({
      showModal: !prevState.showModal,
      eventName,
      selectedId: id
    }));
  };

  updateEventStatus = registeredEvents => {
    const { allEvents } = this.state;
    const newArr = allEvents.map(v => ({
      ...v,
      isSelected: !!registeredEvents.includes(v.id)
    }));

    localStorage.setItem('selectedEvents', JSON.stringify(registeredEvents));

    this.setState({
      selectedEvents: newArr,
      allEvents: newArr
    });
  };

  signUpToEvent = () => {
    const { selectedId, registeredEvents } = this.state;
    registeredEvents.push(selectedId);
    this.updateEventStatus(registeredEvents);
    toastr.success('You have been registered to the event!');
    this.toggleModal();
  };

  cancelEvent = id => {
    const { registeredEvents } = this.state;

    const arrIndex = registeredEvents.indexOf(id);
    registeredEvents.splice(arrIndex, 1);
    this.updateEventStatus(registeredEvents);
    toastr.success('Event cancelled');
  };

  selectEventType = eventType => {
    const { allEvents } = this.state;

    const filteredEvents =
      eventType === EventTypeEnums.MY_EVENTS
        ? allEvents.filter(ev => ev.isSelected === true)
        : allEvents;

    this.setState({
      selectedEvents: filteredEvents
    });
  };

  render() {
    const { selectedEvents, citiesState, eventName, showModal } = this.state;
    return (
      <>
        <Header selectEventType={this.selectEventType} />
        <div className={s.bodyHolder}>
          <Aside filterByName={this.filterByName} filterFreeEvents={this.filterFreeEvents} />
          <div className={s.childrenHolder}>
            <Events
              events={selectedEvents}
              cities={citiesState}
              toggleModal={this.toggleModal}
              signUpToEvent={this.signUpToEvent}
              cancelEvent={this.cancelEvent}
            />

            <Modal show={showModal} handleClose={this.toggleModal}>
              <span>
                Would you like to attend
                <strong>{eventName}</strong>
                event?
              </span>
              <Button onClick={this.signUpToEvent}>Yes</Button>
            </Modal>
          </div>
        </div>
      </>
    );
  }
}

Home.propTypes = {
  events: PropTypes.arrayOf(PropTypes.shape()),
  cities: PropTypes.arrayOf(PropTypes.shape()),
  getEvents: PropTypes.func.isRequired,
  getCities: PropTypes.func.isRequired
};

Home.defaultProps = {
  events: [],
  cities: []
};

const mapStateToProps = createStructuredSelector({
  events: eventsSelector,
  cities: citiesSelector
});

export const mapDispatchToProps = dispatch => ({
  getEvents: () => dispatch(getEvents()),
  getCities: () => dispatch(getCities())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Home);
