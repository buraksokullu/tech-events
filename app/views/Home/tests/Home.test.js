import React from 'react';
import { shallow } from 'enzyme';
import * as trivagoApi from 'Store/api/index';
import { Home, mapDispatchToProps } from '../Home';

jest.mock('Store/api/index', () => ({
  getCampaignData: jest.fn()
}));

describe('Home', () => {
  let homeProps = null;
  const getEventsMock = jest.fn();
  const getCitiesMock = jest.fn();

  const dumEvents = [
    {
      id: 0,
      isFree: true,
      name: 'CSS Grids: fact or fiction',
      city: 9,
      startDate: '2019-07-14T02:00:00+00:00',
      endDate: '2019-07-14T03:00:00+00:00'
    }
  ];

  const dumCities = [
    {
      id: 1,
      name: 'Barcelona'
    }
  ];

  beforeEach(() => {
    homeProps = {
      getEvents: getEventsMock,
      getCities: getCitiesMock,
      events: dumEvents,
      cities: dumCities
    };
  });
  it('Should render correctly', () => {
    const component = shallow(<Home {...homeProps} />);

    expect(component).toMatchSnapshot();
  });
  it('componentDidMount should invoke getEvents and getCities', () => {
    const component = shallow(<Home {...homeProps} />);
    const instance = component.instance();

    instance.getEvents = jest.fn();
    instance.componentDidMount();
    expect(getCitiesMock).toHaveBeenCalled();
    expect(instance.getEvents).toHaveBeenCalled();
  });
  it('getEvents function should get events invoke setData ', async () => {
    const component = shallow(<Home {...homeProps} />);
    const instance = component.instance();

    instance.setData = jest.fn();
    getEventsMock.mockResolvedValue();

    instance.getEvents();

    await expect(
      Promise.resolve().then(async () => {
        expect(getEventsMock).toHaveBeenCalled();
        expect(instance.setData).toHaveBeenCalled();
      })
    );
  });
  it('setData function should get events invoke setData ', async () => {
    const component = shallow(<Home {...homeProps} />);
    const instance = component.instance();

    instance.setState = jest.fn();

    const dumEvents = [
      {
        city: 9,
        endDate: '2019-07-14T03:00:00+00:00',
        id: 0,
        isFree: true,
        isSelected: false,
        name: 'CSS Grids: fact or fiction',
        startDate: '2019-07-14T02:00:00+00:00'
      }
    ];

    const dumCities = [
      {
        id: 1,
        name: 'Barcelona'
      }
    ];
    instance.setData();
    expect(instance.setState).toHaveBeenCalledWith({
      allEvents: dumEvents,
      selectedEvents: dumEvents,
      citiesState: dumCities,
      registeredEvents: []
    });
  });
});
