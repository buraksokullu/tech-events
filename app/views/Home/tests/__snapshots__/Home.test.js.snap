// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Should render correctly 1`] = `
<Fragment>
  <Header
    selectEventType={[Function]}
  />
  <div>
    <Aside
      filterByName={[Function]}
      filterFreeEvents={[Function]}
    />
    <div>
      <Events
        cancelEvent={[Function]}
        cities={Array []}
        events={Array []}
        signUpToEvent={[Function]}
        toggleModal={[Function]}
      />
      <Modal
        handleClose={[Function]}
        show={false}
      >
        <span>
          Would you like to attend
          <strong />
          event?
        </span>
        <Button
          onClick={[Function]}
        >
          Yes
        </Button>
      </Modal>
    </div>
  </div>
</Fragment>
`;
